Mathlib/Tactic/Relation/Trans.lean

```diff
@@ -109,9 +109,31 @@ that is, a relation which has a transitivity lemma tagged with the attribute [tr
 
 * `trans s` replaces the goal with the two subgoals `t ~ s` and `s ~ u`.
 * If `s` is omitted, then a metavariable is used instead.
+
+Additionally, `trans` also applies to a goal whose target has the form `t → u`,
+in which case it replaces the goal with `t → s` and `s → u`.
 -/
 elab "trans" t?:(ppSpace colGt term)? : tactic => withMainContext do
   let tgt ← getMainTarget''
+  match tgt with
+  | .forallE name binderType body info =>
+    -- only consider non-dependent arrows
+    if body.hasLooseBVars then
+      throwError "`trans` is not implemented for dependent arrows{indentExpr tgt}"
+    -- parse the intermeditate term
+    let middleType ← mkFreshExprMVar none
+    let t'? ← t?.mapM (elabTermWithHoles · middleType (← getMainTag))
+    let middle ← (t'?.map (pure ·.1)).getD (mkFreshExprMVar middleType)
+    liftMetaTactic fun goal => do
+      -- create two new goals
+      let g₁ ← mkFreshExprMVar (some <| .forallE name binderType middle info) .synthetic
+      let g₂ ← mkFreshExprMVar (some <| .forallE name middle body info) .synthetic
+      -- close the original goal with `fun x => g₂ (g₁ x)`
+      goal.assign (.lam name binderType (.app g₂ (.app g₁ (.bvar 0))) .default)
+      pure <| [g₁.mvarId!, g₂.mvarId!] ++ if let some (_, gs') := t'? then gs' else [middle.mvarId!]
+    return
+  | _ => pure ()
+
   let (rel, x, z) ←
```

Comment:
I'd be tempted to put the implication handling here instead, by setting `rel` to `(. -> .)`; is that problematic in some way that I'm not seeing?

Reply:
I looked at it and I think there are a few things that would make it impossible or less readable:

- `rel` is later expected to be an `Expr` of the form `.const` (for example in `mkAppM' rel`) and `(· → ·)` would be I guess something like `.lam x _ (.lam y _ (.forallE _ x y _) _) _`, which would require various modifications. I think that could be circumvented by creating `def Impl (a b):= (a → b)`, but I don't think that's a good idea.
- For the same reason, one couldn't put the `trans` attribute on a lemma like `{A B C : Prop} : (A → B) → (B → C) → (A → C)`, so one would need to modify that part to not search for trans lemmas in this case.
- Less importantly, it seems inefficient to search for relevant `trans` lemmas when there is only one thing to do.

However, I realised that the match statement can be moved around, so that it is one single `match tgt with`. This might indirectly address your comment. (so f223943 is big, but most of it is changing the indent)